src/Orleans/Core/GrainFactory.cs

```diff
@@ -85,6 +97,182 @@ public static TGrainInterface GetGrain<TGrainInterface>(string primaryKey, strin
                     typeof(TGrainInterface)));
         }
 
+        /// <summary>
+        /// Creates a reference to the provided <paramref name="obj"/>.
+        /// </summary>
+        /// <typeparam name="TGrainObserverInterface">
+        /// The specific <see cref="IGrainObserver"/> type of <paramref name="obj"/>.
+        /// </typeparam>
+        /// <param name="obj">The object to create a reference to.</param>
+        /// <returns>The reference to <paramref name="obj"/>.</returns>
+        public static Task<TGrainObserverInterface> CreateObjectReference<TGrainObserverInterface>(IGrainObserver obj)
+            where TGrainObserverInterface : IGrainObserver
+        {
+            var interfaceType = typeof(TGrainObserverInterface);
+            if (!interfaceType.IsInterface)
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        "The provided type parameter must be an interface. '{0}' is not an interface.",
+                        interfaceType.FullName));
+            }
+
+            if (!interfaceType.IsInstanceOfType(obj))
+            {
+                throw new ArgumentException(
+                    string.Format("The provided object must implement '{0}'.", interfaceType.FullName),
+                    "obj");
+            }
+            
+            Delegate creator;
+
+            if (!referenceCreators.TryGetValue(interfaceType, out creator))
+            {
+                creator = referenceCreators.GetOrAdd(interfaceType, CreateCreateObjectReferenceDelegate);
+            }
+
+            var resultTask = ((Func<TGrainObserverInterface, Task<TGrainObserverInterface>>)creator)((TGrainObserverInterface)obj);
+            return resultTask;
+        }
+
+        /// <summary>
+        /// Deletes the provided object reference.
+        /// </summary>
+        /// <typeparam name="TGrainObserverInterface">
+        /// The specific <see cref="IGrainObserver"/> type of <paramref name="obj"/>.
+        /// </typeparam>
+        /// <param name="obj">The reference being deleted.</param>
+        /// <returns>A <see cref="Task"/> representing the work performed.</returns>
+        public static Task DeleteObjectReference<TGrainObserverInterface>(
+            IGrainObserver obj) where TGrainObserverInterface : IGrainObserver
```

Comment:
Same here - should/could it be
`public static Task DeleteObjectReference<TGrainObserverInterface>(TGrainObserverInterfaceobj)`
instead?